### ISSUE TITLE
opt: seems like a missing projection after a filter

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#59173]
Consider the following setup:
```
create table t (a int, b int, sum int, s string, primary key (a, b), index bs (b, s));
insert into t select i // 10, i % 10, i//10+i%10, i::string from generate_series(1, 99) as g(i);
```
and query
```
explain (types) select b from t where b > 5 and s like '7%' and sum < 15;
                              info
-----------------------------------------------------------------
  distribution: local
  vectorized: true

  • project
  │ columns: (b int)
  │ estimated row count: 11
  │
  └── • filter
      │ columns: (b int, sum int, s string)
      │ estimated row count: 11
      │ filter: ((sum)[int] < (15)[int])[bool]
      │
      └── • index join
          │ columns: (b int, sum int, s string)
          │ estimated row count: 4
          │ table: t@primary
          │ key columns: a, b
          │
          └── • filter
              │ columns: (a int, b int, s string)
              │ estimated row count: 4
              │ filter: ((s)[string] LIKE ('7%')[string])[bool]
              │
              └── • scan
                    columns: (a int, b int, s string)
                    estimated row count: 40
                    table: t@bs
                    spans: /6/"7"-
(28 rows)
```
I'm a bit surprised to see column `s` propagated after the filter that used it (i.e. I'm expecting for `s` to be pruned right after the "lower" filter), but for some reason it is marked as "needed" for the index join. To me it seems like a projection is missing after `filter`, or am I missing something?

Another interesting observation is that the estimated stats don't make sense after the index joiner - we expect 4 rows coming out of the index join, then we apply filter and somehow we now expect 11 rows, this is quite puzzling.

**Comment by mgartner** [comment]
TLDR; this is a known limitation of the optimizer that is not easy to fix.

During exploration, normalization rules (pruning rules are normalization rules) are only applied to non-root expressions that are created. In order to prune `s` we'd have to apply the pruning normalization rules to the parent of the root expression created during the `GenerateConstrainedScans` exploration rule.

`GenerateConstrainedScans` creates a new `Select` expression (the top-most one below). If pruning normalization rules were applied to this `Select`, there would be nothing to prune because the `Select` cannot know which columns are required by any parent expressions that it is a child of (we only "see" a single parent of each expression, but in a memo forest, a given expression may have many parents). Pruning rules would have to be applied to the top-level `Project` in order to prune `s`.

```
================================================================================
GenerateConstrainedScans
  Cost: 586.60
================================================================================
   project
    ├── columns: b:2!null
    └── select
         ├── columns: b:2!null sum:3!null s:4!null
  -      ├── scan t
  -      │    └── columns: b:2!null sum:3 s:4
  +      ├── index-join t
  +      │    ├── columns: b:2!null sum:3 s:4
  +      │    └── select
  +      │         ├── columns: a:1!null b:2!null s:4!null
  +      │         ├── key: (1,2)
  +      │         ├── fd: (1,2)-->(4)
  +      │         ├── scan t@bs
  +      │         │    ├── columns: a:1!null b:2!null s:4
  +      │         │    ├── constraint: /2/4/1: [/6/'7' - ]
  +      │         │    ├── key: (1,2)
  +      │         │    └── fd: (1,2)-->(4)
  +      │         └── filters
  +      │              └── s:4 LIKE '7%' [outer=(4), constraints=(/4: [/'7' - /'8'); tight)]
         └── filters
  -           ├── b:2 > 5 [outer=(2), constraints=(/2: [/6 - ]; tight)]
  -           ├── s:4 LIKE '7%' [outer=(4), constraints=(/4: [/'7' - /'8'); tight)]
              └── sum:3 < 15 [outer=(3), constraints=(/3: (/NULL - /14]; tight)]
```

**Comment by RaduBerinde** [comment]
The row count of the upper filter is derived through a different expression than the one we see in the final plan (specifically the one where we have the full filter on top of a full-table scan). In an ideal world, the two estimations would reach the same result, but that is pretty hard to achieve in practice.

**Comment by RaduBerinde** [comment]
> I'm a bit surprised to see column s propagated after the filter that used it (i.e. I'm expecting for s to be pruned right after the "lower" filter), but for some reason it is marked as "needed" for the index join. To me it seems like a projection is missing after filter, or am I missing something?

We do a best-effort attempt to prune columns, but there is a tradeoff - adding too many Project operators interferes with various rules from working. We discussed trying to do a final pruning at the end (perhaps using a "soft presentation" physical property), but we haven't seen many cases where this was a real problem.

I will close this for now as there isn't anything immediately actionable.
